gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js

```diff
@@ -0,0 +1,443 @@
+        });
+    };
+
+    gpii.oauth2.selectionTree.setValueAndAllAncestors = function (model, segs, value) {
+        model.value = value;
+        if (segs.length > 0) {
+            // build out the structure if it doesn't exist yet
+            if (!model.children) {
+                model.children = {};
+            }
+            if (!model.children[segs[0]]) {
+                model.children[segs[0]] = {};
+            }
+            // recurse down the segs
+            gpii.oauth2.selectionTree.setValueAndAllAncestors(model.children[segs[0]], segs.slice(1), value);
+        }
+    };
+
+    /**
+     * Sets the "value" property of each ancestor by traversing backwards through the path segments.
+     * The state is calculated by determining the states of all the children.
+     * checked (all children checked), unchecked (all children unchecked), indeterminate (mixture of checked, unchecked, and/or indeterminate)
+     */
+    gpii.oauth2.selectionTree.updateAncestors = function (model, segs) {
+        var parentSegs = fluid.copy(segs);
+        while (parentSegs.length > 0) {
+            parentSegs.pop();
+            var parent = gpii.oauth2.selectionTree.getModelNode(model, parentSegs);
+            parent.value = gpii.oauth2.selectionTree.calculateValueFromChildren(parent);
+        }
+    };
+
+    gpii.oauth2.selectionTree.calculateValueFromChildren = function (model) {
+        var calculatedValue;
+        fluid.each(model.children, function (childModel) {
+            var childValue = childModel.value;
+            if (!calculatedValue) {
+                calculatedValue = childValue;
+            }
+            if (childValue === "indeterminate" || calculatedValue !== childValue) {
+                calculatedValue = "indeterminate";
+            }
+        });
+        return calculatedValue;
+    };
+
+    gpii.oauth2.selectionTree.gatherPaths = function (paths, prefix, model) {
+        var nodeValue = model.value;
+        if (nodeValue === "checked") {
+            paths.push(prefix.join("."));
+        } else if (nodeValue === "indeterminate") {
+            fluid.each(model.children, function (childModel, seg) {
+                prefix.push(seg);
+                gpii.oauth2.selectionTree.gatherPaths(paths, prefix, childModel);
+                prefix.pop();
+            });
+        }
+    };
+
+    gpii.oauth2.selectionTree.toServerModel = function (model) {
+        var paths = [];
+        gpii.oauth2.selectionTree.gatherPaths(paths, [], model);
+        return fluid.arrayToHash(paths);
+    };
+
+    gpii.oauth2.selectionTree.pathsToSegs = function (paths) {
+        var pathsArray = fluid.keys(paths);
+        return fluid.transform(pathsArray, function (elPath) {
+            return fluid.model.pathToSegments(elPath);
+        });
+    };
+
+    gpii.oauth2.selectionTree.toModel = function (setPrefs, clientAvailablePrefs) {
+        var model = {};
```

Comment:
Done in https://github.com/simonbates/universal/commit/b757f6f1de0bcd1b491aee2093c2626b45b5e7d2

Though I used "newSelections" rather than "newModel" as I moved the selections tree in the model to a container called "selections" -- to accommodate the new "hasSelection" boolean value. model.selections and model.hasSelection are siblings. See https://github.com/simonbates/universal/blob/b757f6f1de0bcd1b491aee2093c2626b45b5e7d2/gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js#L42-L60